File: src/endpoints/file/put_item.py
from . import __types__ as types, __libs__ as file_libs
from fastapi import UploadFile
from src import output
from src.libs.db import DB, Table
from src.libs.object import json_parse, json_stringify
from src.libs.string import create_random_string
from src.modules.verify import checking_token
from src.modules.preference import Preference

async def put_item(params: dict = {}, req = None, _db: DB = None, _token = None):

    # set values
    result = None
    db = _db if _db else DB().connect()
    file = {
        'content': None,
        'path': '',
        'name': '',
        'mime': '',
        'ext': '',
    }

    try:
        # set params
        params = types.PutItem(**params)

        # checking token
        token = checking_token(req, db) if not _token else _token

        # check parse json
        json_data = json_parse(params.json_data) if params.json_data else {}

        # read file
        _file: UploadFile = params.file
        file['content'] = await _file.read() if _file else None
        if not file['content']: raise Exception('File not found.', 400)

        # set preference
        pref = Preference()

        # check file size
        if pref.get('file.limitSize') < len(file['content']):
            raise Exception('File size limit exceeded.', 400)

        # check exist module item
        match params.module:
            case 'article': table_name = Table.ARTICLE.value
            case 'json': table_name = Table.JSON.value
            case 'checklist': table_name = Table.CHECKLIST.value
            case 'comment': table_name = Table.COMMENT.value
            case _: table_name = ''
        if not table_name: raise Exception('Module item not found.', 400)
        count = db.get_count(
            table_name=table_name,
            where=[ f'srl = {params.module_srl}' ],
        )
        if not (count > 0): raise Exception('Module item not found.', 400)

        # set file info
        file['name'] = _file.filename
        file['mime'] = _file.content_type
        file['ext'] = file.get('mime').split('/')[1]

        # convert file format
        if file_libs.use_convert_image_format(file.get('mime'), params.file_format):
            file = file_libs.convert_image_format(
                file=file,
                mime=params.file_format,
                quality=params.file_quality,
            )

        # set path
        file['path'] = f'{file_libs.get_dir_path(params.dir_name)}/{file_libs.get_unique_name(8)}.{file['ext']}'

        # copy file
        file_libs.write_file(file['content'], file['path'])

        # set image size
        if file['mime'].startswith('image/'):
            json_data = file_libs.update_image_size(json=json_data, path=file['path'])

        # set values
        values = {
            'name': file['name'],
            'code': create_random_string(16),
            'path': file['path'],
            'mime': file['mime'],
            'size': len(file['content']),
            'json': json_stringify(json_data),
            'module': params.module,
            'module_srl': params.module_srl,
        }

        # set placeholders
        placeholders = [
            { 'key': 'name', 'value': ':name' },
            { 'key': 'code', 'value': ':code' },
            { 'key': 'path', 'value': ':path' },
            { 'key': 'mime', 'value': ':mime' },
            { 'key': 'size', 'value': ':size' },
            { 'key': 'json', 'value': ':json' },
            { 'key': 'module', 'value': ':module' },
            { 'key': 'module_srl', 'value': ':module_srl' },
            { 'key': 'created_at', 'value': 'DATETIME("now", "localtime")' },
        ]

        # add item
        data = db.add_item(
            table_name = Table.FILE.value,
            placeholders = placeholders,
            values = values,
        )

        # get item
        new_item = db.get_item(
            table_name=Table.FILE.value,
            where=[ f'srl = {data}' ],
        )

        # set result
        result = output.success({
            'message': 'Complete add file.',
            'data': {
                'srl': new_item['srl'],
                'code': new_item['code'],
                'name': new_item['name'],
                'mime': new_item['mime'],
                'size': new_item['size'],
                'json': json_parse(new_item['json']) if new_item.get('json') else None,
                'module': new_item['module'],
                'module_srl': new_item['module_srl'],
                'created_at': new_item['created_at'],
            },
        }, _req=req)
    except Exception as e:
        if file['path']: file_libs.delete_file(file['path'])
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/file/put-item.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import { createCode } from '@/libs/strings'
import { deleteFile, getUploadPath } from '@/libs/file'
import { parseJSON } from '@/libs/objects'
import * as helper from './__helper'
import type Service from '@/classes/Service'
import type { FileModel } from './__model'
import type { FileToResource } from './__helper'

type PutItemParams = {
  body: FileModel['putItemBody']
  service: Service
}

export default async function putItem({ body, service }: PutItemParams)
{
  let _path: string = ''
  try
  {
    // check exist origin data
    const originCount = db.getCount({
      table: helper.getModuleName(body.module),
      where: `srl = ${body.module_srl}`,
    })
    if (originCount.data <= 0)
    {
      throw new ServiceError('Not found origin data.', { status: 400 })
    }

    // check file size
    if (body.file.size > service.preference['file.limit.size'])
    {
      throw new ServiceError('File size limit exceeded.', { status: 400 })
    }

    // TODO: 등록할 수 있는 최대 갯수 제한검사

    // convert file to resource
    let _resource: FileToResource = await helper.fileToResource(body.file)

    // convert format or quality
    if (_resource.image)
    {
      _resource.image = helper.convertImageFormat(_resource.image, _resource.mime, body.quality)
    }

    // set save path
    _path = await getUploadPath(body.dir_name, `${createCode(12)}.${_resource.ext}`)

    // copy file
    if (_resource.image)
    {
      await _resource.image.write(_path)
    }
    else if (_resource.bytes)
    {
      await Bun.write(_path, _resource.bytes)
    }

    // set image size
    let _json: ZZ = body.json || {}
    if (_resource.image)
    {
      const _metadata = await _resource.image.metadata()
      _json.width = _metadata.width
      _json.height = _metadata.height
    }

    // add data
    const added = db.addData({
      table: DB.TABLE.FILE,
      values: [
        { key: 'code', value: createCode(8) },
        { key: 'name', value: _resource.name },
        { key: 'path', value: _path },
        { key: 'mime', value: _resource.mime },
        { key: 'size', value: _resource.size },
        { key: 'json', value: JSON.stringify(_json) },
        { key: 'module', value: body.module },
        { key: 'module_srl', value: body.module_srl },
        { key: 'created_at', valueName: DB.DATE_TIME },
      ],
    })

    // get item
    const item = db.getData({
      table: DB.TABLE.FILE,
      where: `srl = ${added.data}`,
    })

    return {
      ...item.data,
      path: undefined,
      json: parseJSON(item.data.json),
    }
  }
  catch (_e: any)
  {
    // delete error file
    if (_path) await deleteFile(_path)
    throw new ServiceError('Failed to add File.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
